src/freertos_drivers/esp32/Esp32Gpio.hxx

```diff
@@ -0,0 +1,467 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * \file Esp32Gpio.hxx
+ *
+ * Helper declarations for using GPIO pins via the ESP-IDF APIs.
+ *
+ * @author Mike Dunston
+ * @date 27 March 2020
+ */
+
+#ifndef _DRIVERS_ESP32GPIO_HXX_
+#define _DRIVERS_ESP32GPIO_HXX_
+
+#include "freertos_drivers/arduino/GpioWrapper.hxx"
+#include "os/Gpio.hxx"
+#include "utils/logging.h"
+#include "utils/macros.h"
+
+#include <driver/gpio.h>
+
+/// Helper macro to test if a pin has been configured for output.
+///
+/// This is necessary since ESP-IDF does not expose gpio_get_direction(pin).
+#define IS_GPIO_OUTPUT(pin) (GPIO_IS_VALID_OUTPUT_GPIO(pin) &&                 \
+                             (pin < 32) ? GPIO.enable & BIT(pin & 31) :        \
+                                          GPIO.enable1.data & BIT(pin & 31))
+
+template <class Defs, bool SAFE_VALUE, bool INVERT> struct GpioOutputPin;
+template <class Defs, bool PUEN, bool PDEN> struct GpioInputPin;
+
+/// Defines a GPIO output pin. Writes to this structure will change the output
+/// level of the pin. Reads will return the pin's current level.
+///
+/// The pin is set to output at initialization time, with the level defined by
+/// `SAFE_VALUE'.
+///
+/// Do not use this class directly. Use @ref GPIO_PIN instead.
+template <gpio_num_t PIN_NUM, bool INVERTED = false>
+class Esp32Gpio : public Gpio
+{
+public:
+#if defined(CONFIG_IDF_TARGET_ESP32S2) || defined(CONFIG_IDF_TARGET_ESP32S3)
+    static_assert(PIN_NUM >= 0 && PIN_NUM <= 46, "Valid pin range is 0..46.");
+    static_assert(!(PIN_NUM >= 22 && PIN_NUM <= 24)
+                , "Pin does not exist");
+    static_assert(!(PIN_NUM >= 26 && PIN_NUM <= 32)
+                , "Pin is reserved for flash usage.");
+#else
+    static_assert(PIN_NUM >= 0 && PIN_NUM <= 39, "Valid pin range is 0..39.");
+    static_assert(PIN_NUM != 24, "Pin does not exist");
+    static_assert(!(PIN_NUM >= 28 && PIN_NUM <= 31), "Pin does not exist");
+    static_assert(PIN_NUM != 37, "Pin is connected to GPIO 36 via capacitor.");
+    static_assert(PIN_NUM != 38, "Pin is connected to GPIO 39 via capacitor.");
+#if defined(ESP32_PICO)
+    static_assert(!(PIN_NUM >= 6 && PIN_NUM <= 8)
+                , "Pin is reserved for flash usage.");
+    static_assert(PIN_NUM != 11 && PIN_NUM != 16 && PIN_NUM != 17
+                , "Pin is reserved for flash usage.");
+#else
+    static_assert(!(PIN_NUM >= 6 && PIN_NUM <= 11)
+                , "Pin is reserved for flash usage.");
+#if defined(BOARD_HAS_PSRAM)
+    static_assert(PIN_NUM != 16 && PIN_NUM != 17
+                , "Pin is reserved for PSRAM usage.");
+#endif // BOARD_HAS_PSRAM
+#endif // ESP32_PICO
```

Comment:
I find these a bit confusing, because really the block this is closing is the !PICO

Reply:
The complexity comes from the PICO-D4 allowing usage of pins 9 and 10 whereas all other variants restrict those pins to flash. I'll see if I can clarify this further in comments.